File: lib/test/t_UnionMesher_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class UnionMesher
 *
 *  Copyright 2005-2026 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include <iostream>

#include "otmeshing/otmeshing.hxx"
#include "openturns/OTtestcode.hxx"
#include "openturns/IntervalMesher.hxx"

using namespace OT;
using namespace OTMESHING;
using namespace OT::Test;

typedef UnionMesher::MeshCollection MeshCollection;

int main()
{
  TESTPREAMBLE;

  // 1. Default construction and string representation
  UnionMesher mesher;
  std::cout << mesher << std::endl;

  // 2. Empty collection → empty mesh (dimension 0)
  {
    Mesh empty(mesher.build(MeshCollection()));
    std::cout << "Empty collection: " << empty << std::endl;
    assert_equal(empty.getDimension(), static_cast<UnsignedInteger>(0));
    assert_equal(empty.getVerticesNumber(), static_cast<UnsignedInteger>(0));
  }

  // 3. Single mesh → returns compressed mesh (no-op for valid input)
  for (UnsignedInteger dim = 2; dim < 5; ++dim)
  {
    const Mesh mesh(IntervalMesher(Indices(dim, 1)).build(Interval(dim)));
    const Mesh single(mesher.build(MeshCollection(1, mesh)));
    std::cout << "Single dim=" << dim << ": " << single << std::endl;
    assert_equal(single.getVerticesNumber(), mesh.getVerticesNumber());
    assert_equal(single.getSimplicesNumber(), mesh.getSimplicesNumber());
    assert_almost_equal(single.getVolume(), 1.0);
  }

  // 4. Non-overlapping meshes
  for (UnsignedInteger dim = 2; dim < 6; ++dim)
  {
    const Mesh mesh1(IntervalMesher(Indices(dim, 1)).build(Interval(dim)));
    const Mesh mesh2(IntervalMesher(Indices(dim, 1)).build(Interval(Point(dim, 2.0), Point(dim, 3.0))));
    const Mesh unionMesh(mesher.build(MeshCollection({mesh1, mesh2})));
    std::cout << "Disjoint dim=" << dim << ": " << unionMesh << std::endl;
    assert_almost_equal(unionMesh.getVolume(), 2.0);
  }

  // 5. Touching meshes (shared boundary) — exercises vertex dedup
  // mesh1 = [0,1]^dim, mesh2 = [1,2]x[0,1]^{dim-1}, sharing the face at x=1
  for (UnsignedInteger dim = 2; dim < 5; ++dim)
  {
    const Mesh mesh1(IntervalMesher(Indices(dim, 1)).build(Interval(dim)));
    Point lower(dim, 0.0);
    Point upper(dim, 1.0);
    lower[0] = 1.0;
    upper[0] = 2.0;
    const Mesh mesh2(IntervalMesher(Indices(dim, 1)).build(Interval(lower, upper)));
    const Mesh unionMesh(mesher.build(MeshCollection({mesh1, mesh2})));
    std::cout << "Touching dim=" << dim << ": " << unionMesh << std::endl;
    assert_almost_equal(unionMesh.getVolume(), 2.0);
    // shared (dim-1)-face has 2^{dim-1} vertices
    const UnsignedInteger nNoDedup = (1u << (dim + 1));
    const UnsignedInteger nShared = (1u << (dim - 1));
    assert_equal(unionMesh.getVerticesNumber(), nNoDedup - nShared);
  }

  // 6. Three disjoint meshes
  for (UnsignedInteger dim = 2; dim < 4; ++dim)
  {
    const Mesh mesh1(IntervalMesher(Indices(dim, 1)).build(Interval(dim)));
    const Mesh mesh2(IntervalMesher(Indices(dim, 1)).build(Interval(Point(dim, 2.0), Point(dim, 3.0))));
    const Mesh mesh3(IntervalMesher(Indices(dim, 1)).build(Interval(Point(dim, 4.0), Point(dim, 5.0))));
    const Mesh unionMesh(mesher.build(MeshCollection({mesh1, mesh2, mesh3})));
    std::cout << "Three dim=" << dim << ": " << unionMesh << std::endl;
    assert_almost_equal(unionMesh.getVolume(), 3.0);
  }

  // 7. CompressMesh static method: no-op on clean mesh
  {
    const Mesh mesh(IntervalMesher(Indices(2, 1)).build(Interval(2)));
    const Mesh compressed(UnionMesher::CompressMesh(mesh));
    std::cout << "Already compressed: " << compressed << std::endl;
    assert_equal(compressed.getVerticesNumber(), mesh.getVerticesNumber());
    assert_equal(compressed.getSimplicesNumber(), mesh.getSimplicesNumber());
    assert_almost_equal(compressed.getVolume(), 1.0);
  }

  // 8. CompressMesh: collapse duplicate vertices
  {
    Sample vertices(0, 2);
    vertices.add(Point({0.0, 0.0}));
    vertices.add(Point({1.0, 0.0}));
    vertices.add(Point({1.0, 1.0}));
    vertices.add(Point({0.0, 1.0}));
    vertices.add(Point({0.0, 0.0}));
    vertices.add(Point({1.0, 0.0}));
    IndicesCollection simplices(4, 3);
    simplices(0, 0) = 0; simplices(0, 1) = 1; simplices(0, 2) = 2;
    simplices(1, 0) = 0; simplices(1, 1) = 2; simplices(1, 2) = 3;
    simplices(2, 0) = 4; simplices(2, 1) = 5; simplices(2, 2) = 2;
    simplices(3, 0) = 4; simplices(3, 1) = 2; simplices(3, 2) = 3;
    Mesh mesh(vertices, simplices, false);
    const Mesh compressed(UnionMesher::CompressMesh(mesh));
    std::cout << "Duplicates collapsed: " << compressed << std::endl;
    assert_equal(compressed.getVerticesNumber(), static_cast<UnsignedInteger>(4));
    assert_almost_equal(compressed.getVolume(), 2.0);
  }

  // 9. CompressMesh: unused vertices dropped
  {
    Sample vertices(0, 2);
    vertices.add(Point({0.0, 0.0}));
    vertices.add(Point({1.0, 0.0}));
    vertices.add(Point({1.0, 1.0}));
    vertices.add(Point({0.0, 1.0}));
    vertices.add(Point({42.0, 42.0}));
    IndicesCollection simplices(2, 3);
    simplices(0, 0) = 0; simplices(0, 1) = 1; simplices(0, 2) = 2;
    simplices(1, 0) = 0; simplices(1, 1) = 2; simplices(1, 2) = 3;
    Mesh mesh(vertices, simplices, false);
    const Mesh compressed(UnionMesher::CompressMesh(mesh));
    std::cout << "Unused dropped: " << compressed << std::endl;
    assert_equal(compressed.getVerticesNumber(), static_cast<UnsignedInteger>(4));
    assert_almost_equal(compressed.getVolume(), 1.0);
  }

  // 10. build with a mesh that has an unused vertex
  {
    Sample vertices(0, 2);
    vertices.add(Point({0.0, 0.0}));
    vertices.add(Point({1.0, 0.0}));
    vertices.add(Point({1.0, 1.0}));
    vertices.add(Point({0.0, 1.0}));
    vertices.add(Point({42.0, 42.0}));
    IndicesCollection simplices(2, 3);
    simplices(0, 0) = 0; simplices(0, 1) = 1; simplices(0, 2) = 2;
    simplices(1, 0) = 0; simplices(1, 1) = 2; simplices(1, 2) = 3;
    Mesh mesh(vertices, simplices, false);
    const Mesh unionMesh(mesher.build(MeshCollection({mesh})));
    std::cout << "Build with unused: " << unionMesh << std::endl;
    assert_equal(unionMesh.getVerticesNumber(), static_cast<UnsignedInteger>(4));
    assert_almost_equal(unionMesh.getVolume(), 1.0);
  }

  // 11. Empty input mesh
  {
    const Mesh emptyMesh(Sample(0, 2), IndicesCollection());
    const Mesh single(mesher.build(MeshCollection({emptyMesh})));
    std::cout << "Empty mesh: " << single << std::endl;
  }

  // 12. Two empty meshes
  {
    const Mesh emptyMesh(Sample(0, 2), IndicesCollection());
    const Mesh unionMesh(mesher.build(MeshCollection({emptyMesh, emptyMesh})));
    std::cout << "Two empty: " << unionMesh << std::endl;
  }

  // 13. Dimension mismatch should throw
  {
    const Mesh mesh1(IntervalMesher(Indices(2, 1)).build(Interval(2)));
    const Mesh mesh2(IntervalMesher(Indices(3, 1)).build(Interval(3)));
    try
    {
      mesher.build(MeshCollection({mesh1, mesh2}));
      throw TestFailed(OSS() << "Expected InvalidArgumentException");
    }
    catch (const InvalidArgumentException &)
    {
      std::cout << "Dimension mismatch correctly raised" << std::endl;
    }
  }

  // 14. CompressMesh: duplicate vertices with non-zero range (tolerance > 0)
  {
    Sample vertices(0, 2);
    vertices.add(Point({0.0, 0.0}));
    vertices.add(Point({1.0, 0.0}));
    vertices.add(Point({1.0, 1.0}));
    vertices.add(Point({0.0, 1.0}));
    vertices.add(Point({0.0, 0.0}));
    vertices.add(Point({1.0, 0.0}));
    IndicesCollection simplices(4, 3);
    simplices(0, 0) = 0; simplices(0, 1) = 1; simplices(0, 2) = 2;
    simplices(1, 0) = 0; simplices(1, 1) = 2; simplices(1, 2) = 3;
    simplices(2, 0) = 4; simplices(2, 1) = 5; simplices(2, 2) = 2;
    simplices(3, 0) = 4; simplices(3, 1) = 2; simplices(3, 2) = 3;
    Mesh mesh(vertices, simplices, false);
    const Mesh compressed(UnionMesher::CompressMesh(mesh));
    std::cout << "Duplicates non-zero range: " << compressed << std::endl;
    assert_equal(compressed.getVerticesNumber(), static_cast<UnsignedInteger>(4));
    assert_almost_equal(compressed.getVolume(), 2.0);
  }

  return 0;
}

File: python/test/t_UnionMesher_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import otmeshing

ot.TESTPREAMBLE()

mesher = otmeshing.UnionMesher()
print(mesher)
print(repr(mesher))

# 1. Empty collection -> empty mesh
empty = mesher.build([])
print(f"Empty collection: {empty}")
ott.assert_almost_equal(empty.getDimension(), 0)
ott.assert_almost_equal(empty.getVerticesNumber(), 0)

# 2. Single mesh -> returns compressed mesh (no-op for valid input)
for dim in range(2, 5):
    mesh = ot.IntervalMesher([1] * dim).build(ot.Interval(dim))
    single = mesher.build([mesh])
    print(f"Single dim={dim}: {single}")
    assert single.isValid()
    ott.assert_almost_equal(single.getVerticesNumber(), mesh.getVerticesNumber())
    ott.assert_almost_equal(single.getSimplicesNumber(), mesh.getSimplicesNumber())
    ott.assert_almost_equal(single.getVolume(), 1.0)

# 3. Non-overlapping meshes (original test, extended)
for dim in range(2, 6):
    mesh1 = ot.IntervalMesher([1] * dim).build(ot.Interval(dim))
    mesh2 = ot.IntervalMesher([1] * dim).build(ot.Interval([2.0] * dim, [3.0] * dim))
    union = mesher.build([mesh1, mesh2])
    print(f"Disjoint dim={dim}: {union}")
    assert union.isValid()
    ott.assert_almost_equal(union.getVolume(), 2.0)

# 4. Touching meshes (shared boundary) -- exercises vertex dedup
# mesh1 = [0,1]^dim, mesh2 = [1,2]x[0,1]^{dim-1}, sharing the face at x=1
for dim in range(2, 5):
    mesh1 = ot.IntervalMesher([1] * dim).build(ot.Interval(dim))  # [0,1]^dim
    lower = [1.0] + [0.0] * (dim - 1)
    upper = [2.0] + [1.0] * (dim - 1)
    mesh2 = ot.IntervalMesher([1] * dim).build(ot.Interval(lower, upper))
    union = mesher.build([mesh1, mesh2])
    print(f"Touching dim={dim}: {union}")
    assert union.isValid()
    ott.assert_almost_equal(union.getVolume(), 2.0)
    # shared (dim-1)-face has 2^{dim-1} vertices
    nVertices = union.getVerticesNumber()
    nNoDedup = 2 ** (dim + 1)
    nShared = 2 ** (dim - 1)
    expectedVertices = nNoDedup - nShared
    print(f"  vertices={nVertices} expected={expectedVertices}")
    ott.assert_almost_equal(nVertices, expectedVertices)

# 5. Three disjoint meshes combined
for dim in range(2, 4):
    mesh1 = ot.IntervalMesher([1] * dim).build(ot.Interval(dim))
    mesh2 = ot.IntervalMesher([1] * dim).build(ot.Interval([2.0] * dim, [3.0] * dim))
    mesh3 = ot.IntervalMesher([1] * dim).build(ot.Interval([4.0] * dim, [5.0] * dim))
    union = mesher.build([mesh1, mesh2, mesh3])
    print(f"Three dim={dim}: {union}")
    assert union.isValid()
    ott.assert_almost_equal(union.getVolume(), 3.0)

# 6. CompressMesh static method: duplicate vertices
mesh = ot.IntervalMesher([1] * 2).build(ot.Interval(2))
compressed = otmeshing.UnionMesher.CompressMesh(mesh)
print(f"Already compressed: {compressed}")
assert compressed.isValid()
ott.assert_almost_equal(compressed.getVerticesNumber(), mesh.getVerticesNumber())
ott.assert_almost_equal(compressed.getSimplicesNumber(), mesh.getSimplicesNumber())
ott.assert_almost_equal(compressed.getVolume(), 1.0)

# 7. CompressMesh static method: collapse duplicates
vertices = ot.Sample(
    [
        [0.0, 0.0],
        [1.0, 0.0],
        [1.0, 1.0],
        [0.0, 1.0],
        [0.0, 0.0],
        [1.0, 0.0],
        [1.0, 1.0],
        [0.0, 1.0],
    ]
)
simplices = ot.IndicesCollection([[0, 1, 2], [0, 2, 3], [4, 5, 6], [4, 6, 7]])
mesh = ot.Mesh(vertices, simplices, False)
compressed = otmeshing.UnionMesher.CompressMesh(mesh)
print(f"Duplicates collapsed: {compressed}")
assert compressed.isValid()
ott.assert_almost_equal(compressed.getVerticesNumber(), 4)
ott.assert_almost_equal(compressed.getVolume(), 2.0)

# 8. CompressMesh static method: used-vertex-only output
# A mesh where only some vertices are referenced by simplices
vertices = ot.Sample(
    [[0.0, 0.0], [1.0, 0.0], [1.0, 1.0], [0.0, 1.0], [42.0, 42.0]]
)  # unused vertex
simplices = ot.IndicesCollection([[0, 1, 2], [0, 2, 3]])
mesh = ot.Mesh(vertices, simplices, False)
compressed = otmeshing.UnionMesher.CompressMesh(mesh)
print(f"Unused dropped: {compressed}")
assert compressed.isValid()
ott.assert_almost_equal(compressed.getVerticesNumber(), 4)
ott.assert_almost_equal(compressed.getVolume(), 1.0)

# 9. build with a mesh that has an unused vertex
union = mesher.build([mesh])
print(f"Build with unused: {union}")
assert union.isValid()
ott.assert_almost_equal(union.getVerticesNumber(), 4)
ott.assert_almost_equal(union.getVolume(), 1.0)

# 10. Empty input mesh
empty_mesh = ot.Mesh(ot.Sample(0, 2), ot.IndicesCollection())
single = mesher.build([empty_mesh])
print(f"Empty mesh: {single}")
assert single.isValid()

# 11. Two empty meshes
union = mesher.build([empty_mesh, empty_mesh])
print(f"Two empty: {union}")
assert union.isValid()

# 12. Dimension mismatch should raise
mesh1 = ot.IntervalMesher([1] * 2).build(ot.Interval(2))
mesh2 = ot.IntervalMesher([1] * 3).build(ot.Interval(3))
with ott.assert_raises(TypeError):
    mesher.build([mesh1, mesh2])
print("Dimension mismatch correctly raised")

# 13. CompressMesh: duplicate vertices with non-zero range (tolerance > 0)
# Mix of unique and duplicate vertices so tolerance is non-zero
pts = [
    [0.0, 0.0],
    [1.0, 0.0],
    [1.0, 1.0],
    [0.0, 1.0],
    [0.0, 0.0],
    [1.0, 0.0],
]  # last two are duplicates of first two
simplices = ot.IndicesCollection([[0, 1, 2], [0, 2, 3], [4, 5, 2], [4, 2, 3]])
vertices = ot.Sample(pts)
mesh = ot.Mesh(vertices, simplices, False)
compressed = otmeshing.UnionMesher.CompressMesh(mesh)
print(f"Duplicates with non-zero range: {compressed}")
assert compressed.isValid()
# 4 unique vertices + 2 duplicates -> 4 compressed vertices
ott.assert_almost_equal(compressed.getVerticesNumber(), 4)
ott.assert_almost_equal(compressed.getVolume(), 2.0)
